File: server/src/test/java/de/zalando/zally/rules/LimitNumberOfSubresourcesRuleTest.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import io.swagger.models.Swagger;
import io.swagger.parser.SwaggerParser;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

public class LimitNumberOfSubresourcesRuleTest {

    @Test
    public void shouldReturnViolationIfReachLimit() {
        Swagger swagger = getFixture("fixtures/limitNumberOfSubresourcesInvalid.json");
        Rule rule = new LimitNumberOfSubresourcesRule();
        List<Violation> violations = rule.validate(swagger);
        assertEquals(1, violations.size());
    }

    @Test
    public void shouldNotReturnViolations() {
        Swagger swagger = getFixture("fixtures/limitNumberOfSubresourcesValid.json");
        Rule rule = new LimitNumberOfSubresourcesRule();
        List<Violation> violations = rule.validate(swagger);
        assertEquals(0, violations.size());
    }

    private Swagger getFixture(String fixture) {
        return new SwaggerParser().read(fixture);
    }
}

File: server/src/test/java/de/zalando/zally/rules/LimitNumberOfSubresourcesRuleTest.kt
package de.zalando.zally.rules

import de.zalando.zally.getFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class LimitNumberOfSubresourcesRuleTest {
    @Test
    fun positiveCase() {
        val swagger = getFixture("limitNumberOfSubresourcesValid.json")
        assertThat(LimitNumberOfSubresourcesRule().validate(swagger)).isNull()
    }

    @Test
    fun negativeCase() {
        val swagger = getFixture("limitNumberOfSubresourcesInvalid.json")
        assertThat(LimitNumberOfSubresourcesRule().validate(swagger)!!.paths).hasSameElementsAs(
                listOf("/items/{some_id}/item_level_1/{level1_id}/item-level-2/{level2_id}/item-level-3/{level3_id}/item-level-4")
        )
    }

    @Test
    fun positiveCaseSpp() {
        val swagger = getFixture("api_spp.json")
        assertThat(AvoidLinkHeadersRule().validate(swagger)).isNull()
    }

    @Test
    fun positiveCaseSpa() {
        val swagger = getFixture("api_spa.yaml")
        assertThat(AvoidLinkHeadersRule().validate(swagger)).isNull()
    }
}
